src/checker.rs

```diff
@@ -0,0 +1,88 @@
+use std::error::Error;
+use std::time::Duration;
+
+use reqwest::{ClientBuilder, StatusCode};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum CheckResult {
+    SUCCESS,
+    FAILURE(FailureReason),
+    MISSED,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+pub enum FailureReason {
+    Timeout,
+    DnsError(String),
+    ConfigError(String),
+    Error(String),
+}
+
+pub async fn check_domain(domain: String) -> CheckResult {
+    // TODO: Timeout should come from config
+    let timeout = Duration::new(5, 0);
+    let client_result = ClientBuilder::new().timeout(timeout).build();
+    let client = match client_result {
+        Err(e) => {
+            return CheckResult::FAILURE(FailureReason::ConfigError(e.to_string()));
+        }
+        Ok(client) => client,
+    };
+    let mut response_result = client.head(&domain).send().await;
+    if let Ok(response) = &response_result {
+        if response.status() == StatusCode::METHOD_NOT_ALLOWED {
+            // If this domain doesn't accept HEAD, attempt GET request
+            response_result = client.get(&domain).send().await;
+        }
+    }
```

Comment:
Immediately after this you're matching on `response_result` anyway. You can move this into the `Ok` branch.

Reply:
Oh, sorry, I didn't catch that `response_result` was mutable. That works, but in my opinion it's a bit surprising. If you had
1. `CheckResult` as a conventional `Result` type (see above),
2. An impl of `From<reqwest::Error> for `CheckError`
this could look like
```rust
    let response = client.head(&domain).send().await?;
    if response.status() == StatusCode::METHOD_NOT_ALLOWED {
        client.get(&domain).send().await?;
    }
    Ok(())

Reply:
This is discarding too much right. I think we would want

```rs
let response = client.head(&domain).send().await?;
match response.status() {
  StatusCode::METHOD_NOT_ALLOWED => client.get(&domain).send().await?
  _ => response
}
```